### ISSUE TITLE
Error 137 Build Failures

### ISSUE THREAD

**Issue by C-STYR** [gravitational/next#789]
Several deploys to preview (and potentially merges into `main` - unclear) have failed with `exit code 137`.

There are several [here](https://vercel.com/gravitational/teleport-docs-next) for analysis. Redeploying is usually successful the second time. 

It's worth noting that we have a Pro-level account with Vercel with 12 concurrent cloud build slots (the maximum available for purchase).  Because our deployments are sometimes queued it is clear that we are exceeding this level of use between /next, /teleport, and other repos.  While this may not be directly related to the problem, it may be an indication that we are exceeding capacity on other infrastructure related to code 137.

More info on that [here](https://vercel.com/support/articles/why-are-my-vercel-builds-queued).

**Comment by wadells** [comment]
Some quick googling points to oomkiller, which terminates processes with 137.  I think we should open a ticket with Vercel support to discuss this.  They don't mention any memory limits for builds on [their limits page](https://vercel.com/docs/concepts/limits/overview). That indicates a platform failure to me -- they shouldn't be killing our jobs for using too much memory unless they document those limits.

The approach from our end would be optimising the build to use less memory, but we should talk to Vercel first to figure out if this behavior is expected, and what limit we'd need to stay under.

**Comment by C-STYR** [comment]
@iAdramelk can you please have a look at this when you're next in and see what can be done on our side?

**Comment by iAdramelk** [comment]
@C-STYR ok I run some tests locally with different settings and checked memory usage. Its hard to exactly measure memory usage in this case, but I have some rough numbers and results.

TLDR, it's the same reason as with the serverless functions  error that @Alqanar fixed. Merging her PR should fix this error for now. Also, merging back PR with mdx update and esm will reduce memory usage by 15% more percents.

Long story. Our bottleneck with memory is the `webpack` build process. And there are two things that affecting it most:

1) Amount of the entries (*js-files inside the `/pages/` folder) that `webpack` needs to process.
2) Amount and complexity of computation `webpack` and other parts of `next` do on the bundles generated from these files.

OoM error in our case is a combination of both. First we have 1000+ entries in the `/pages/` folder, which is a lot, but should not give us OoM error by itself, but here the second reason comes along. Starting from `next@12` they have this option enabled by default https://nextjs.org/docs/advanced-features/output-file-tracing . Then this option is enabled, next build will try to statically analyze not only client js files, but also server side js files to create more efficient server side build (copy only files that are actually used on srerver, instead of full content of `node_modules`). And combination of these two things breaks the build. Disabling this flag fixes build for now.

Some very rough measurements of the memory use on my locally:

- `next@12` build with `outputFileTracing` enabled - up to 4,71Gb of memory.
- `next@12` build with `outputFileTracing` disabled - up to 2,89Gb of memory.
- With `outputFileTracing` disabled, but if also merge all blog pages - up to 3,41Gb of memory.
- Same as before, but if we also merge `esm` PR with more efficient `mdx` compiler – up to 2,90Gb of memory.

Some observations here:

- `outputFileTracing` increases memory requirement up to 2 times.
- More entry points (! pages, its two different things) to `webpack` bundle increses memory.
- More efficient compilation of some page parts (e. g.) `mdx` reduces memory. (Now I'm interested how removing `styled-components` will affect build time)

If we disable `outputFileTracing` and update `mdx` memory usage will return to under 3Gb which is fine for now, but will became the problem again after we add 500+ more pages. Just splitting the repos will only temporarily solve the issue, at some point of time were will be more that 2000K pages in the docs and build will break again. To solve the problem once and for all we need to reduce the number of entry points to the `webpack` bundle without changing the number of generated pages. I have some ideas about how to do that and I will write separate proposal for it.

If we don't have new OoM error, we can close this issue in couple of days.

**Comment by alexwolfe** [comment]
Thanks. 
- Can you provide memory usage if we remove the blog and website from the docs from the repo
- Can you provide an estimate % based on how much more memory styled components uses vs css. You could create a set of dummy pages not using the site and compare them building 100 pages

**Comment by iAdramelk** [comment]
@alexwolfe without blog and docs it is up to `1,09Gb`.

For build without `styled-component` it is hard to estimate. Dummy pages will not help here, because memory usage is affected mostly not by the number of pages itself, but by the size of import graph (e. g. page imports 20 components, each imports 20 more files, and they import also 10 libraries each with 100 files). So such test will return super rough values.
